src/charm.py

```diff
@@ -460,27 +460,68 @@ def _iscsi_initiator(self, tenv: Environment) -> None:
         if initiators:
-            # search for hostname and create context
+            # search for hostname in initiator-dictionary if it exists
             initiators_dict = json.loads(initiators)
             if hostname in initiators_dict.keys():
                 initiator_name = initiators_dict[hostname]
 
-        if not initiator_name:
+        initiator_name_from_file = self._get_initiator_name_from_file(self.ISCSI_INITIATOR_NAME)
+
+        # either initiator-dictionary not provided or hostname not present there
+        # and config file doesn't contain initiator name
+        if not initiator_name and not initiator_name_from_file:
+            # generate random iqn
             initiator_name = subprocess.getoutput("/sbin/iscsi-iname")
             logging.warning(
-                "The hostname was not found in the initiator dictionary! "
-                "The random iqn %s will be used for %s",
+                "Hostname was not found in initiator-dict and config file. "
+                + "The randomly generated iqn %s will be used for %s",
                 initiator_name,
                 hostname,
             )
+            render_file = True
 
-        logging.info("Rendering initiatorname.iscsi")
-        ctxt = {"initiator_name": initiator_name}
-        template = tenv.get_template("initiatorname.iscsi.j2")
-        rendered_content = template.render(ctxt)
-        self.ISCSI_INITIATOR_NAME.write_text(rendered_content)
+        if initiator_name and initiator_name != initiator_name_from_file:
+            render_file = True
+
+        if render_file:
+            logging.info("Rendering initiatorname.iscsi")
+            ctxt = {"initiator_name": initiator_name}
+            template = tenv.get_template("initiatorname.iscsi.j2")
+            rendered_content = template.render(ctxt)
+            self.ISCSI_INITIATOR_NAME.write_text(rendered_content)
+
+    def _get_initiator_name_from_file(self, iscsi_config_file: Path | None = None) -> str | None:
+        """Parse iSCSI config file to check if "InitiatorName" is set.
+
+        iSCSI config file is /etc/iscsi/initiatorname.iscsi by default.
+
+        Returns:
+            "InitiatorName" value in config file if present.
+            None if config file isn't present or "InitiatorName" isn't set.
+        """
+        # can't set default param value to instance attribute
+        if not iscsi_config_file:
+            iscsi_config_file = self.ISCSI_INITIATOR_NAME
+        try:
+            with open(iscsi_config_file, encoding="utf-8") as file:
+                lines = file.readlines()
+        except FileNotFoundError:
+            logging.warning("%s doesn't exist", iscsi_config_file)
+            return None
+
+        initiator_name, value = None, None
+        for line in lines:
+            line = line.strip()
+            if line and not line.startswith("#"):
+                initiator_name, value = line.split("=")
```

Comment:
Is that possible we have multiple names in the file? What happend if there is multiple line?

Reply:
Usually the file will only contain something like this:

```
# comment 1
# comment 2
initiatorName = some-iqn
```

Ideally, there shouldn't be multiple names in this file since it's supposed to only contain one definition of the `initiatorName` key. But in case there are more entries like so:

```
# comment 1
# comment 2
initiatorName = some-iqn
initiatorName = another-iqn
```

Then `initiator_name` will be `initiatorName` and `value` would be `another-iqn` which is the last entry.

Reply:
I've gone ahead with the implementation given below so we can be more explicit about our intention to pick the first key, value pair after the comment (which is usually the case since we'd only have one "key = value" pair).

```python
        initiator_name, value = None, None
        for line in lines:
            line = line.strip()
            if line and not line.startswith("#"):
                initiator_name, value = line.split("=")
                break
```

I've added a `break` statement so that after finding the first instance of `initiatorName`, it will exit the loop.